text/trace/0205-messaging-semantic-conventions-context-propagation.md

```diff
@@ -0,0 +1,173 @@
+coordinated way. Future versions of these conventions might recommend [context propagation according to certain industry standards](#standards-for-context-propagation).
+
+### Requirements
+
+A producer SHOULD attach a creation context to each message. The creation context
+SHOULD be attached in a way so that it is not possible to be changed by intermediaries.
+
+## Future possibilities
+
+### Transport context propagation
+
+As illustrated in the [example](#example) above, a creation context can be
+attached to a message, while different contexts are propagated via HTTP requests
+publishing and fetching the message. When coming up with conventions and
+guidance for intermediary instrumentation, it will be beneficial to clearly
+outline those two layers of context propagation and build conventions for
+intermediary instrumentation on top of this outline:
+
+1. The _creation context layer_ allows correlating the producer with the
+   consumers of a message, regardless of intermediary instrumentation. The
+   creation context is created by the producer and must be propagated to the
+   consumers. It must not be altered by intermediaries.
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. An additional _transport context layer_ allows correlating the producer and
+   the consumer with an intermediary. It also allows correlating multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries that
+   are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+This would keep the existing correlation between producers and consumers intact
+while allowing intermediaries to use the transport context to correlate
+intermediary instrumentation with existing producer and consumer
+instrumentations.
+
+### Standards for context propagation
+
+Currently, instrumentors have to decide how to attach and extract context from
+messages to fulfil the [requirements for context propagation](#context-propagation).
+While preserving the freedom for instrumentors to choose how to propagate
+context, in the future these conventions should list recommended ways of how to
+propagate context using popular messaging protocols.
+
+Currently, several attempts exist to standardize context propagation for different
+messaging protocols and scenarios:
+
+* [AMQP](https://w3c.github.io/trace-context-amqp/)
+* [MQTT](https://w3c.github.io/trace-context-mqtt/)
+* [CloudEvents via HTTP](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md)
+
+Those standards are in draft states and/or are not widely adopted yet. It is
+planned to drive those standards to a stable state and to make sure they cover
```

Comment:
This part is a bit confusing to me. I think it's trying to say 2 things in the same paragraph. 

>  It is planned to drive those standards to a stable state and to make sure they cover requirements put forth by these semantic conventions

Who is exactly involved in "planned to drive those standards"? Is it us - OTel?  If so, then maybe making it more clear like:

> OpenTelemetry is going to help drive these standards, making sure they are compatible and cover the requirements put forth by these semantic conventions

And: 

> Finally, these semantic conventions should give a clear and stable recommendation for each protocol and scenario

I think this is a bit redundant? The paragraph above `in the future these conventions should list recommended ways..` says the same thing. I guess we can still keep this as a more "strong" assertion, but rephrase it a bit and make it a separated statement from the above:

> Once the standards reach a stable state, these semantic conventions will give a clear and stable recommendation for each protocol and scenario.

WDYT?

>

Reply:
> Who is exactly involved in "planned to drive those standards"? Is it us - OTel?

No, those standards are outside OTel. However, people working on those are counting on OTel to validate whether the specification the came up with makes sense and can be stable.

I agree this paragraph can be confusing. I will simplify it and just go with the last suggestion you made:

> Once the standards reach a stable state, these semantic conventions will give a clear and stable recommendation for each protocol and scenario.

Reply:
BTW, I know we're not defining the standards, and having them defined would take a lot of time. Can we still have some soft recommendations of what people should do in absence of those?

E.g. "while the standards are being defined, we suggest that instrumentations that follow this convention use [Propagator API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#propagators-api) to propagate context using default or user-configured propagator."

Maybe we can also discuss it at the Thu meeting - I think we need to understand how it can evolve after AMQP/MQTT and other standards come into play and new instrumentations will have to stay backward-compatible.

Reply:
@lmolkova We can discuss this on Thursday.

From my point of view, this PR makes requirements for context propagation clear, but it doesn't give any recommendations for context propagation on top of that. As at this point we cannot give any stable recommendations (as currently none of the standard messaging protocols has a stable specification for context propagation established), I think we should refrain from giving any recommendations.

Also, it would be hard to give generic recommendations, as those recommendations would always be specific to certain protocols and messaging systems.

However, I'm open to other suggestions here, I'll put it on the agenda for Thursday.

Reply:
Summarizing the discussion in today's messaging workgroup. Giving clear recommendations for context propagation requires 3 sequential steps:
1. Making requirements for context propagation clear (this PR).
2. Verifying that protocols (AMQP, MQTT, CloudEvents) satisfy those requirements, if so, drive context propagation extensions for those protocols to a stable state.
3. Recommend stable context propagation mechanisms for those protocols in OpenTelemetry semantic conventions.

Before all three steps are done, we will not really be able to give stable recommendations for context propagation, and users will be left with a certain (intended) uncertainty here.

Reply:
Created https://github.com/w3c/trace-context-binary/issues/19 on the W3C binary protocol to prevent potential header name collisions between formats